crates/bevy_ecs/src/query/access.rs

```diff
@@ -49,17 +49,27 @@ impl<'a, T: SparseSetIndex + fmt::Debug> fmt::Debug for FormattedBitSet<'a, T> {
 /// See the [`is_compatible`](Access::is_compatible) and [`get_conflicts`](Access::get_conflicts) functions.
 #[derive(Eq, PartialEq)]
 pub struct Access<T: SparseSetIndex> {
-    /// All accessed elements.
-    reads_and_writes: FixedBitSet,
-    /// The exclusively-accessed elements.
-    writes: FixedBitSet,
-    /// Is `true` if this has access to all elements in the collection.
+    /// All accessed components.
+    component_read_and_writes: FixedBitSet,
+    /// The exclusively-accessed components.
+    component_writes: FixedBitSet,
+    /// All accessed resources.
+    resource_read_and_writes: FixedBitSet,
```

Comment:
Why introduce extra FixedBitSets for resources here?  IMO, we only need to split 'reads/writes_all' into 'read/write_all_component/resource' because Components and Resources share ID space

Reply:
It's to check the `has_any_resource` that is needed to check compatibility

Reply:
There's also a correctness hazard in sharing the bit sets for resources and components. There's a risk that the component Foo clashes with the resource Foo, for types which are both components and resources. This isn't the right memory model, and care must be taken far elsewhere in the stack (when assigning component ids) to avoid it if we don't separate it here.

A test to validate that behavior would be nice actually.

Reply:
Not for this PR, but we should discuss whether to split the ID space. With this PR each bitset is going to be oversized and do extra work for the id's of the other type.

Reply:
what I am most worried about is that `Access ` is used in many places (such as EntityFilterRef), Perhaps we can move it to a new struct `CombinedAccess ` and used it for' FilteredAccessSet::combined_access, with hhis way, we can limit the influence of this pr.

Reply:
> Not for this PR, but we should discuss whether to split the ID space. With this PR each bitset is going to be oversized and do extra work for the id's of the other type.

Note that this probably won't be a problem anymore if we go ahead with this
https://github.com/bevyengine/bevy/pull/14385
which removes the bitsets.
(which will probably be necessary since with relations the ComponentIds will become very high, so the memory used by FixedBitSet will become prohibitive)

> what I am most worried about is that Access  is used in many places (such as EntityFilterRef), Perhaps we can move it to a new struct CombinedAccess  and used it for' FilteredAccessSet::combined_access, with hhis way, we can limit the influence of this pr.

Your concern is the increased memory usage because of 2 FixedBitSets?
Yes i think the solution would probably be https://github.com/bevyengine/bevy/pull/14385

Reply:
> A test to validate that behavior would be nice actually.

@alice-i-cecile added a test that checks that a resource/component that share the same componentId (same struct) do not conflict with each other

Reply:
> yeah ,Two FixedBitSets (even not used) occupy extra 48 bytes, we only use it for combined access,
With https://github.com/bevyengine/bevy/pull/14385, this number is even Larger

On the contrary with https://github.com/bevyengine/bevy/pull/14385 we don't have a problem anymore, no?
14385 uses a `Vec<usize>` for the access; so if there are no resource accessed we would just have an empty vec.

Or do you mean stack usage and not heap usage?

Reply:
with #14385 It will make the problem more serious.
1. `Access` is just a struct, whether it occupies stack meormory or heap moery depends on how you use it
2. #14385  uses SmallVec<[usize;N> (N default to 64),so it occupies 24+8*64 =704 bytes even without any usage.